### PR TITLE
Allow the user to delete his own account.

### PR DESCRIPTION
**Pull request by Leo-MouraF**
Hey boss! The taks is done. I think have some issues to solve, but we can talk later. xoxo.

### CODE REVIEW

**Comment by guiUrbanas03** [comment]
@Leo-MouraF Congratulations for concluding the task, the account deletion is working as expected! There are just some details that I noticed that we could make it better. Please, check the comments that I left in the code on the "Files changed" section.

File: source/blueprints/user/services.py [hunk 1]
```diff
@@ -130,4 +131,23 @@ def verify_password(current_password: str, new_password: str, confirm_password:
...
 
-    return hash_password(new_password)
\ No newline at end of file
+    return hash_password(new_password)
+
+
+def delete_user(answer: str):
+    """Save the date of delete account at column 'deleted_at' for User, Profile and Game Stats."""
+    
+    if answer == "yes":
```
Comment (guiUrbanas03) [inline source/blueprints/user/services.py:140]: The logic of receiving a "confirm_deletion" flag is correct, but we don't need to have it in the backend, in this context this is a concern of the frontend, so we can securely remove it here.

File: source/blueprints/user/routes.py [hunk 2]
```diff
@@ -121,3 +122,31 @@ def change_password():
...
+        db.session.commit()
+
+    except ValueError as error:
+        db.session.rollback()
+        abort(422, {"type": CauseTypeError.VALIDATION_ERROR.value, "data": str(error)})
+        
+    return DataResponse(
+        "User deletion succesfully",
+        200,
+        {
```
Comment (guiUrbanas03) [inline source/blueprints/user/routes.py:145]: Just for a matter of keeping the patterns, in the data response dict replace the keys "User", "Profile", "Game Stats" with "user", "profile" and "game_stats", respectively.

**Review by guiUrbanas03 — CHANGES_REQUESTED** [review]

**Comment by Leo-MouraF** [comment]
Hey boss. I've already done the changes that you asked me. I think you'll find a problem. I'm pasting the code here, so this ways you can see exactly what i mean.
def user_del():
    """Allow the user to delete his own account"""

    try:
        if not current_user.deleted_at:
            delete_user()
            db.session.commit()
            
        else:   
            return 'User already deleted.'

So, here I imagine that we don't need the "else", but when i tried to make run without this else, didn't run correctly.

File: source/blueprints/user/routes.py [hunk 3]
```diff
@@ -128,12 +128,13 @@ def change_password():
...
     try:
-        delete_confirm = UserDeleteDTO().load(data)
-        delete_user(delete_confirm["confirm_deletion"])
-        db.session.commit()
+        if not current_user.deleted_at:
+            delete_user()
+            db.session.commit()
+            
+        else:   
+            return 'User already deleted.'
```
Comment (guiUrbanas03) [inline source/blueprints/user/routes.py:137]: Just a small tweak here,  replace `return 'User already deleted.'` with the following code:

```
abort(409, {"type": CauseTypeError.DATA_VIOLATION_ERROR.value, "data": "User already deleted."})
```

Note that I choose the **409 status code**  - which stands for **CONFLICT**.

In this app, usually when we are sending a **successful response** we are going to use the `DataResponse` class, and when we are sending a **error response** we are going to use the `abort()` function.

Let me know if you have any questions about it.

**Review by guiUrbanas03 — CHANGES_REQUESTED** [review]

**Comment by guiUrbanas03** [comment]
> Hey boss. I've already done the changes that you asked me. I think you'll find a problem. I'm pasting the code here, so this ways you can see exactly what i mean. def user_del(): """Allow the user to delete his own account"""
> 
> ```
> try:
>     if not current_user.deleted_at:
>         delete_user()
>         db.session.commit()
>         
>     else:   
>         return 'User already deleted.'
> ```
> 
> So, here I imagine that we don't need the "else", but when i tried to make run without this else, didn't run correctly.

In this situation we really need to have the else statement because the code inside the if block doesn't return, raise or break the flow when a user that is already deleted tries to delete himself again.

But there's something that we could do better in regards the "User already deleted" message, please refer to this comment to see the details:
https://github.com/guiUrbanas03/fk-royale-backend/pull/5#discussion_r1306378261

File: source/blueprints/user/routes.py [hunk 4]
```diff
@@ -121,3 +122,32 @@ def change_password():
         },
     ).json()
 
+
+@user_bp.route("/delete_account", methods=["DELETE"]) 
+@jwt_required()
+def user_del():
```
Comment (guiUrbanas03) [inline source/blueprints/user/routes.py:128]: Juuust another really small tweak here too, prefer to use descriptive names for the functions:
`def delete_user()` it's likely cleaner and more readable than names such as `user_del` or `del_user`.

**Review by guiUrbanas03 — CHANGES_REQUESTED** [review]

**Pull request merged**